client/src/main/java/io/confluent/kafka/schemaregistry/client/CachedSchemaRegistryClient.java

```diff
@@ -243,17 +291,44 @@ protected ParsedSchema getSchemaByIdFromRegistry(int id, String subject)
 
   private int getVersionFromRegistry(String subject, ParsedSchema schema, boolean normalize)
       throws IOException, RestClientException {
-    io.confluent.kafka.schemaregistry.client.rest.entities.Schema response =
-        restService.lookUpSubjectVersion(schema.canonicalString(),
-            schema.schemaType(), schema.references(), subject, normalize, true);
+    if (missingSchemaCache.getIfPresent(new SubjectAndSchema(subject, schema)) != null) {
+      throw new RestClientException("This ID is banned",
+          HTTP_NOT_FOUND, SCHEMA_NOT_FOUND_ERROR_CODE);
+    }
+
+    io.confluent.kafka.schemaregistry.client.rest.entities.Schema response;
+    try {
+      response = restService.lookUpSubjectVersion(schema.canonicalString(),
+              schema.schemaType(), schema.references(), subject, normalize, true);
+    } catch (RestClientException rce) {
+      if (rce.getStatus() == HTTP_NOT_FOUND
+          && rce.getErrorCode() == SCHEMA_NOT_FOUND_ERROR_CODE) {
+        missingSchemaCache.put(new SubjectAndSchema(subject, schema), System.currentTimeMillis());
+      }
+      throw rce;
+    }
+
     return response.getVersion();
   }
 
   private int getIdFromRegistry(String subject, ParsedSchema schema, boolean normalize)
       throws IOException, RestClientException {
-    io.confluent.kafka.schemaregistry.client.rest.entities.Schema response =
-        restService.lookUpSubjectVersion(schema.canonicalString(),
-            schema.schemaType(), schema.references(), subject, normalize, false);
+    if (missingSchemaCache.getIfPresent(new SubjectAndSchema(subject, schema)) != null) {
```

Comment:
Please refactor this code into a separate method since it is duplicated with above.